core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -1900,16 +1902,28 @@ private static BucketToPartition createBucketToPartitionMap(
                 return new BucketToPartition(Optional.of(IntStream.range(0, hashPartitionCount).toArray()), Optional.empty());
             }
             else if (partitioningHandle.getConnectorId().isPresent()) {
-                int partitionCount = hashPartitionCount;
                 BucketNodeMap bucketNodeMap = nodePartitioningManager.getBucketNodeMap(session, partitioningHandle, true);
-                if (!bucketNodeMap.isDynamic()) {
-                    partitionCount = bucketNodeMap.getNodeCount();
-                }
                 int bucketCount = bucketNodeMap.getBucketCount();
                 int[] bucketToPartition = new int[bucketCount];
-                int nextPartitionId = 0;
-                for (int bucket = 0; bucket < bucketCount; bucket++) {
-                    bucketToPartition[bucket] = nextPartitionId++ % partitionCount;
+                if (bucketNodeMap.isDynamic()) {
+                    int nextPartitionId = 0;
+                    for (int bucket = 0; bucket < bucketCount; bucket++) {
+                        bucketToPartition[bucket] = nextPartitionId++ % hashPartitionCount;
+                    }
+                }
+                else {
+                    BiMap<InternalNode, Integer> nodeToPartition = HashBiMap.create();
```

Comment:
I think it' beneficial to add a comment explaining the logic, e.g. "make sure all buckets mapped to the same node map to the same partition, such that locality requirements are respected in scheduling".